src/playground-service-worker-proxy.ts

```diff
@@ -52,73 +52,122 @@ import {
-    'message',
-    (event: MessageEvent<PlaygroundMessage>) => {
-      if (event.data.type === UPDATE_SERVICE_WORKER) {
-        // When the project handshakes with the service worker, it may notice a
-        // version mismatch, in which case it will send this message to request
-        // the service worker update.
-        //
-        // Calling update triggers a fetch of the service worker. If the scripts
-        // are byte-different, then it will be installed + activated, and we'll
-        // receive the "updatefound" even there. Note this is exactly the normal
-        // Service Worker update lifecycle, it's just that calling update()
-        // forces it to happen immediately, as opposed to at some unknown
-        // opportunistic time in the future.
-        //
-        // TODO(aomarks) It would be good to show an error to the user somehow
-        // if we never end up getting a new version here. That would indicate
-        // that the SW and project are out of date *on the server*.
-        registration.update();
-        // There seems to be a bug in Chrome where the previous version of a
-        // service worker sometimes gets stuck and never shuts down when an
-        // updating is pending, even across a full browser restart. Calling
-        // unregister here seems to reliably force the previous version to shut
-        // down.
-        registration.unregister();
-      }
+  /** https://www.w3.org/TR/service-workers/#get-newest-worker-algorithm */
+  const getNewestWorker = () =>
+    registration.installing ?? registration.waiting ?? registration.active;
+
+  /** The service worker we most recently connected to. */
+  let newest: ServiceWorker | null = null;
+
+  /**
+   * Resolve when the given service worker reaches its activated state.
+   */
+  const activated = async (sw: ServiceWorker): Promise<void> => {
+    if (sw.state === 'activated') {
+      return;
     }
-  );
+    return new Promise<void>((resolve) => {
+      const aborter = new AbortController();
+      sw.addEventListener(
+        'statechange',
+        () => {
+          if (sw.state === 'activated') {
+            resolve();
+            aborter.abort();
+          }
+        },
+        {signal: aborter.signal}
+      );
+    });
+  };
 
-  const connect = (sw: ServiceWorker) => {
-    const {port1, port2} = new MessageChannel();
+  const connectToNewest = async () => {
+    const sw = getNewestWorker();
+    newest = sw;
+    if (sw === null) {
+      // Can't find a way to reproduce this scenario, possibly can't happen.
+      console.error('No playground service worker found.');
+      return;
+    }
 
+    // Don't connect to the project until the service worker is activated.
+    // Otherwise the project might try to load a preview before the service
+    // worker is actively controlling the preview's URL space.
+    await activated(sw);
+    if (sw !== newest) {
+      // A new service worker appeared while we were waiting.
```

Comment:
It would have to be something very strange where version 0 started loading, and then a new version 1 was published to the server and the registration updated (maybe via another tab, or directly by the user in dev tools) somehow before version 0 finished installing.

Very good chance this would never ever happen in practice.